### PR TITLE
evm: 63/64 gas clamp for ERC20 calls; improved VM error surfacing; add composite Chainlink-like oracle

### PR DESCRIPTION
**Pull request by Unique-Divine**
# Summary

### CI & tooling

* **golangci-lint**: bump GitHub Action from `v2.1` → `v2.4`; `just lint` now parameterizes the Docker image/tag and cache dir with `image_version="v2.4.0"`.
* **Dependencies**: move `github.com/cosmos/cosmos-sdk` replace to `v0.47.11-nibiru.4` (from `.3`); `go.sum` updated accordingly.
* **Hardhat**: enable optimizer (runs: 5) and `viaIR: true` for `solc 0.8.24`.

### Test logging infrastructure

* New `x/common/testutil/log.go`: `LogRoutingSuite` routes `log` package output into `testing.T` with file\:line attribution; suites now embed this to unify log reporting.

### New Solidity: composite Chainlink-like oracle

* **`NibiruOracleChainLinkLikeMulti.sol`**: Chainlink AggregatorV3-compatible read-only feed that composes up to 4 Nibiru oracle pairs by multiplying legs (18-dec inputs, scaled to constructor-specified decimals).

  * Guards: `NegativeOracle`, `ZeroOracle`.
  * Aggregation: `roundId/answeredInRound` take max; `startedAt/updatedAt` take min; final answer truncated to target decimals.

### Malicious/ERC20 test contracts

* Normalize constructors (drop unused `decimals_`); formatting cleanups.
* **`TestInfiniteRecursionERC20.sol`** greatly expanded:

  * Detailed comments on recursion vectors.
  * `balanceOf` re-enters precompile via `staticcall(balance)`.
  * `transfer` and **new** `transferFrom` re-enter via `sendToBank`.
  * Helpers `attackBalance` (view) and `attackTransfer` (tx).

### Gas forwarding & recursion hardening (EIP-150)

* Rename & clarify helper: `getCallGasWithLimit` → **`getCallGasLimit63_64`**, applied across all ERC20 call/query paths.
* **New unit test** `erc20_unit_test.go` verifies 63/64 math, capping, infinite meter, and boundary cases.

### EVM execution & error handling

* **`evm.go`**:

  * New **`HandleOutOfGasPanic`** (captures `sdk.ErrorOutOfGas`; preserves existing err; wraps with format).
  * New **`SafeConsumeGas`**: consumes gas with recovery; returns structured error on panic.
  * **`NewRevertError`** message simplified: `"execution reverted"` (drops “with no reason”).
* **`keeper/call_contract.go`**:

  * Remove `defer HandleOutOfGasPanic(...)` here; switch to applying msg, recording **`ctx.WithLastErrApplyEvmMsg`**, then consuming gas via `evm.SafeConsumeGas`.
  * On failure, enrich VM error: append `LastErrApplyEvmMsg` detail; tailor OOG message to `"VMError: <vmErr>: gas required exceeds gas limit (%d)"`.
* **`keeper/msg_server.go`**:

  * Similar error-flow changes: distinguish consensus vs execution errors from `ApplyEvmMsg`, record last error, consume gas safely, and append `LastErrApplyEvmMsg` into `VmError` if failed.
* **`keeper/apply` path tweaks**:

  * `ApplyEvmMsg`: rearranged intrinsic gas check and error to a direct `fmt.Errorf("ApplyEvmMsg: %s: provided msg.Gas ...", core.ErrIntrinsicGas, ...)`.
  * Parse wei earlier; clearer failure messages.

### Tx construction helpers & tests

* Refactor helpers into a unified **`NewMsgEthereumTx(ArgsEthTx)`** (create/execute); internal creators renamed to `msgEthTxCreateContract` / `msgEthereumTxExecuteContract`.
* Switch execute path to **`gethcore.LegacyTx`** (from AccessList).
* Tests updated accordingly; several assertions strengthened; additional funding step for fee collector to allow second run.

### Precompile plumbing

* `precompile/funtoken.go` & `precompile/wasm.go`: rename `evm` param → `evmObj`; drop local OOG panic handlers (centralize in EVM layer); gas accounting uses `evmObj.Config.Tracer`. Imports reordered.

### gRPC / estimates

* `EstimateGasForEvmCallType`: adjust OOG string (see questions below about wording).

### FunToken tests

* Recursion tests now assert bounded failure behavior (63/64) and better surfaced intrinsic gas errors; improved messages and ordering; one log message clarified.

### CODE REVIEW

File: x/evm/keeper/msg_server.go [hunk 1]
```diff
@@ -66,18 +67,33 @@ func (k *Keeper) EthereumTx(
...
+		if evmResp == nil {
+			// Consensus error - return immediately, skipping the
+			// "evm.SafeConsumeGas" call we do for
+			ctx.WithLastErrApplyEvmMsg(applyErr)
+			return nil, sdkioerrors.Wrap(applyErr, "consensus error in ethereum message")
+		} else {
+			// Execution error - log but continue processing
+			ctx.WithLastErrApplyEvmMsg(applyErr)
+		}
 	}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/keeper/msg_server.go:90]: Not necessary. That function is an in-place setter.

File: x/evm/keeper/msg_server.go [hunk 2]
```diff
@@ -66,18 +67,33 @@
...
-	if err != nil {
-		return nil, sdkioerrors.Wrap(err, "error applying ethereum core message")
+
+	if evmResp != nil {
+		gasErr := evm.SafeConsumeGas(ctx, evmResp.GasUsed, "execute EthereumTx")
+		if gasErr != nil {
+			return nil, gasErr
+		}
 	}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/keeper/msg_server.go:97]: What's there is correct

File: x/evm/keeper/msg_ethereum_tx_test.go [hunk 3]
```diff
@@ -85,13 +86,14 @@ func (s *Suite) TestMsgEthereumTx_CreateContract() {
...
-				}
-				ethTxMsg, err := evmtest.CreateContractMsgEthereumTx(args)
+				ethTxMsg, err := evmtest.NewMsgEthereumTx(evmtest.ArgsEthTx{
+					CreateContract: &evmtest.ArgsCreateContract{
+						EthAcc:        ethAcc,
+						EthChainIDInt: deps.EvmKeeper.EthChainID(deps.Ctx),
+						GasPrice:      big.NewInt(1),
+						Nonce:         deps.NewStateDB().GetNonce(ethAcc.EthAddr),
+					},
+				})
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/keeper/msg_ethereum_tx_test.go:96]: Fixed in 0d5d9965c29f6ddd0c12c99e1b9c5b7170ee82ee

File: x/evm/keeper/erc20_unit_test.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+import (
+	"github.com/stretchr/testify/suite"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type UnitSuite struct {
+	suite.Suite
+}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/keeper/erc20_unit_test.go]: Fixed in 0d5d9965c29f6ddd0c12c99e1b9c5b7170ee82ee

File: x/evm/evm.go [hunk 5]
```diff
@@ -111,3 +113,51 @@ func ValidateFunTokenBankMetadata(
...
+			default:
+				// Non-OOG panics are not handled here
+				panic(perr)
+			}
+		}
+		if errp != nil && *errp != nil && format != "" {
+			*errp = fmt.Errorf("%s: %w", format, *errp)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/evm.go]: Fixed in 0d5d9965c29f6ddd0c12c99e1b9c5b7170ee82ee

File: x/evm/embeds/contracts/TestInfiniteRecursionERC20.sol [hunk 6]
```diff
@@ -4,24 +4,112 @@ pragma solidity ^0.8.0;
...
+        // (erc20=this, account=who). Force recursion via `balanceOf`.
+        (bool ok, bytes memory ret) = address(FUNTOKEN_PRECOMPILE_ADDRESS)
+            .staticcall(
+                abi.encodeWithSignature(
+                    "balance(address,address)",
+                    // IFunToken.balance(who address, funtoken address)
+                    who,
+                    address(this)
+                )
+            );
```
[the remark this answers is not in the document]
Comment (Unique-Divine) [reply x/evm/embeds/contracts/TestInfiniteRecursionERC20.sol:76]: The correct function signature is this 

```solidity
    /// @notice Method "balance" returns the ERC20 balance and Bank Coin balance
    /// of some fungible token held by the given account.
    function balance(
        address who,
        address funtoken // ERC20
    )

```

Thus, what's in the code now is correct.
